### ISSUE TITLE
Error using JoinElementSets when try to merge two models

### ISSUE THREAD

**Issue by YChang20** [gibbonCode/GIBBON#69]
Dear Kevin,

I want to model the sphere indentation on CNCHelix/MatrixBox. I have referenced your demo in runTetGen: 
Example 3: Meshing a two-region model
Example 5: Meshing regions with holes

and I write the Yin_CNCmodel.m as attached.
My plan is to import the densecellulose.stl (attached in the .zip), then create a box.
Merge the face and  vertices, estimate the volume and region, and finally runTetGen.

However, I have the error when the code runs to joinElementSets:

Error using cellfun
Input #2 expected to be a cell array, was double instead.

Error in joinElementSets (line 31)
numVertSets=cellfun(@(x) size(x,1),Vc);

Error in getInnerPoint (line 37)
    [F,V,C]=joinElementSets(F,V);

I have checked I can create the mesh with the two models separately, but do have problems when trying to merge them.
Could you give me some suggestions on solving this?

Thank you. I have attached the .m and the .stl in the attched .zip

Best regards,
Yin
[YinCNCindent.zip](https://github.com/gibbonCode/GIBBON/files/5015055/YinCNCindent.zip)

**Comment by Kevin-Mattheus-Moerman** [comment]
Hi @YChang20 thanks for posting this issue. 

Note that joining only joins the faces/nodes as a set and that merging would force nodes that coincide to become the same node. So merging and joining are two different things and only for merging are the two shapes actually attached to each other with shared nodes. Also, for merging, the nodes have to already coincide for it to work. If you do not intend to merge them you could always mesh the two (with TetGen) separately and then join the elements sets. 

I'll look into this in more detail next week when I have more time. For the moment perhaps my comments above help. It would also help if you could explain in more detail what you are trying to do.

**Comment by YChang20** [comment]
Dear Kevin,

Thank you for the suggestions. My previous method is to generate a box containing all fibers inside. Because the two models are overlapped, I think to merge them is the correct one. However, I still cannot solve the merge problem.
Therefore, I have changed to another way that using Blender software to create a box with boolean substraction the fiber space.
Then try to tetgen two models separately. This method works, and below is my way to solve it (also might be a possible answer to issue #52 ). 

Related to issue #52:
Boundary is not the problem, but the imported model is with bad mesh, so I first use tetgen -R to reconstruct the mesh and out put 4 new files: densematrix.3 .node, .ele, .face, .edge
And modify my matlab code as follow:

fid=fopen('densematrix.3.ele','rt');
fmt=repmat('%f',1,5);
Emat=textscan(fid,fmt,'delimiter','\t', 'HeaderLines', 1);
Emat(:,1)=[];
Emat=cell2mat(Emat);
%The nodes (vertices)
fid=fopen('densematrix.3.node','rt');
fmt=repmat('%f',1,4);
Vmat=textscan(fid,fmt,'delimiter','\t', 'HeaderLines', 1); 
Vmat(:,1)=[];
Vmat=cell2mat(Vmat);
%The boundary faces
fid=fopen('densematrix.3.face','rt');
fmt=repmat('%f',1,5);
Fmat=textscan(fid,fmt,'delimiter','\t', 'HeaderLines', 1); %The nodes (vertices)
Fmat(:,1)=[];
Cmat=Fmat(:,4); %The "colors" or labels for the boundary faces
Cmat=cell2mat(Cmat);
Fmat(:,4)=[];
Fmat=cell2mat(Fmat);
elementMaterialIndicesMat=ones(size(Emat,1),1); %Element material indices

%join two models:
[Fb1,V1,Cb1]=joinElementSets({Fmat,Fcel},{Vmat,Vcel}); %Join sets

The model is successfully generated as attached
![CNCmat](https://user-images.githubusercontent.com/22903956/89405187-1252fb00-d713-11ea-8bd1-7ac3ecab416f.JPG)

It is trivial, so I still hope to know the possible ways to do merge models. 

Thank you.

**Comment by YChang20** [comment]
Dear Kevin,

I found the merge problems also come from the original .stl file that the mesh of initial model is with self-intersectional faces.
I am now able to generate merged models with no problem, but when it runs febio, the model initalization fail due to negative jocobian of the initial model. (I have checked this failure on convergent is not because the indent force)

I am now setting the CNCfiber and cubic matrix with the same material parameters. However, it still cannot converge at the very beginning. Therefore, I suspect the two models still not merge quite good at the interface or the nodes are too close maybe?

Sorry I have tried many ways to find a better mesh method, but still not finding one really general and easy way (I use Blender software and it is easy to create models with self-intersection faces). I was wondering if you could help to look at this new one and any suggestion on the mesh and model is welcome.

Thank you.

[YinCNCindent.zip](https://github.com/gibbonCode/GIBBON/files/5043109/YinCNCindent.zip)
Yin

**Comment by Kevin-Mattheus-Moerman** [comment]
@YChang20 is this problem still remaining? If so let me know. Perhaps we can have a Zoom/Skype session to have a look at this.

**Comment by YChang20** [comment]
Dear Kevin

It is temporary solved for the model I am using now. It is a problem of 
import stl file.
I created models in Blender and it is common tetgen reported the error 
of self intersection of faces.
I then tried to remesh stl models in Blender first. With certain mesh 
size (also cannot be too small) then import in GIBBON code the tetgen 
just works, but I have no idea why.
Because I don't think the mesh in Blender should affect how tetgen mesh 
the imported structure?

In fact in my project, I need to generate fibrous layers and stack them 
helicoidally with different rotation angles. I believe I should be able 
to generate these models in GIBBON directly.
I explored the GIBBON function of rotating the model which can be 
useful. But I cannot find a GIBBON function to generate a cylindrical 
fiber and repeat it to form an array.
These are actually the functions I need to generate and adjust my 
models.
I was wondering if there will be a function like this in GIBBON in the 
future, or if you have some guidance for user to generate new functions 
work in GIBBON?


Thank you for your kind reply. I will be happy to have a zoom and 
discuss about it.

Best,
Yin

**Comment by YChang20** [comment]
Dear Kevin,

Regarding to my previous reply that I found GIBBON 'pachcylinder' and 'tform' are the two functions I need.
I sucessfully produced cylinders embedded in a box and did the tetgen.

I think I found the issue of tetgen the imported Blender model. As the attached figure, I found it always generates self-intersecting facets with the Blender model (circled blue). However, if I use 'pachcylinder' to generate the cylinder with the same geometry, it doesn't have any problem and can be successfully tetgen. However, I think I need your help on Boolean the model with GIBBON code which I will open another issue.

I will close the issue now.
Thank you for the help.
![blender](https://user-images.githubusercontent.com/22903956/99157852-a34b5200-26c4-11eb-8121-6c46c86f4cb6.JPG)
![pachcylinder](https://user-images.githubusercontent.com/22903956/99157996-df32e700-26c5-11eb-9a85-1fc055e0aee0.JPG)
